### ISSUE TITLE
[RFC] Ui layouting and builder

### ISSUE THREAD

**Issue by jojolepro** [amethyst/amethyst#589]
- [X] 
```rust
// Y,X naming
enum Anchor{
	TopLeft,
	TopMiddle,
	TopRight,
	MiddleLeft,
	Middle,
	MiddleRight,
	BottomLeft,
	BottomMiddle,
	BottomRight,
}
enum Stretch{
	X,
	Y,
	Both,
}

// Relative to parent
struct Anchored{
	anchor: Anchor,
	offset: Option<(f32,f32)>, // Defaults to none. If it is none, set to some using UiTransform position
}
impl Component for Anchored

AnchoredSystem{
	(UiTransform,Anchored)

	if anchor.offset = None{
		anchor.offset = transform.offset
	}
}

struct Stretched{
	stretch: Stretch,
	margin: (f32,f32) //default to 0,0; in builder use .with_margin
}
impl Component for Stretched

```

- [ ] 
```rust
// Ui file format is just a serialize of the different builders data.
// Ui manual format is by manually creating entities in the world with the proper components.
```

- [X] 
```rust
//Example button builder
struct ButtonBuilder{
	name: &str,
	posx: f32, //Offset it there's a parent
	posy: f32,
	anchor: Option<Anchored>,
	stretched: Option<Stretched>,
}
```

- [X] 
For all of this to work, the UiTransform should indicate the MIDDLE of the ui element!!!!!!!!!

- [X] 
```rust
//Button sample structure, generated by the button builder, or macro btn!
ButtonEntity
Comps:[
	UiTransform,
	Anchored,
	Stretched,
]
-TextEntity as child
-Comps:[
	UiTransform,
	UiText,
	Anchored,
]
```

- [ ] 
```rust
// UI Builder
// The macro will generate a list of entities.
// It uses the builder pattern for each ui element
// DECLARATIVE SYNTAX
!buildui{
	Button("btn1",0,0).with_anchor(LeftTop),
	DropDownList([elements],20,20).stretched(Stretch::X)
}
Where Button would get converted to
Image(x,y,default_w,default_h,default_button_texture){ // {} == child and Z -= 1
  Text(text).alignX
}

```

**Comment by Xaeroxe** [comment]
@jojolepro We could recycle [this](https://www.amethyst.rs/doc/develop/doc/amethyst_core/transform/components/struct.Parent.html) component.  Or we could make a `UiParent` component that is very similar.

UiTransform being at the center is pretty easy.  You just change the rendering to render the element at (-element width/2, -element height/2)

**Comment by Kerollmops** [comment]
Just thinking about that, but why not reusing [the cassowary crate](https://github.com/dylanede/cassowary-rs) that [the limn crate](https://github.com/christolliday/limn) is actually using ?

**Comment by jojolepro** [comment]
Now is a bit late since the PR with the layouting code is in the review phase. I didn't think about checking if layouting crates already existed, but I figured what we needed it for was simple enough to not require overcomplex solutions.
@Xaeroxe thoughts?

**Comment by Xaeroxe** [comment]
@Aceeri tried to get this working with our stuff for a while to little effect.  I haven't worked with cassowary myself though so maybe he could provide a more concrete opinion on it.

**Comment by AlisCode** [comment]
Regarding your latest thought, using a dedicated crate seems to me like a better idea than implementing everything by hand.

@jojolepro (sorry if this isn't the best place to ask), I'd like to become a contributor. Is there any UI-related issue that you might need a hand with ? (Asking since #577 seems to be taken by another wannabe-contributor).

**Comment by jojolepro** [comment]
@AlisCode 
You could try to implement relative positionning and margins on this PR if you wish. https://github.com/jojolepro/amethyst/tree/uilayout
The data is already there, but the logic isn't. Check layout.rs and UiTransform.
Also, if you want to discuss with faster response time, gitter is always opened on my second screen: https://gitter.im/amethyst/general

**Comment by jojolepro** [comment]
Closed as #607 is more complete.
